content/blog/2022-04-28-chlorophyll/index.md

```diff
@@ -0,0 +1,309 @@
+3. <u>Program synthesis is hard to scale to large programs.</u>
+
+Therefore, the authors propose a synthesis-aided compiler Chlorophyll to solve the above challenges. 
+
+## Methods
+
+In this section, I will talk about the four stages of the compiler, including program partition, layout and routing, code separation, and code generation.
+
+### Program Partition
+Partitioning the program is the most important step in the Chlorophyll. The following subsections will explain the programming model and the partioning process.
+
+#### Programming Model
+Chlorophyll provides an easy-to-use programming model for partitioning. Basically, users can directly write C code and add partition annotations to the program. `@` specifies the variable or the computation is on which partition, and `!` means the data is sent from one partition to another.
+
+Programmers only need to write partially annotated code as shown below, and the compiler will automatically generate the fully annotated one. This example shows `x` is in the `2`nd partition, and `y` is in the `1`st partition. Partition `2` needs to firstly transfer data `x` to partition `1`, and then compute `x * y`. Finally, the result will be passed to partition `0` and returned.
+```cpp
+// partial annotation
+int@0 mult(int x, int y) { return x * y; } 
+// full annotation
+int@0 mult(int@2 x, int@1 y)
+  { return (x!1 *@1 y)!0; }
+```
+
+For control flow, Chlorophyll only supports constant-bound `for` loops and general `while` loops. It will duplicate the control flow in each core to ensure the program's correctness.
+
+For distributed arrays, users can annotate which part of data should be put on which partition, and use `place` to specify the partition based on the data location in the array. For example, the following figure shows `[0:32]` elements in array `x` are put on partition `0`, and `[32:64]` elements are put on partition `1`.
+
+<img src="distributed-arrays.png" alt="alt_text" title="image_tooltip" style="zoom:18%;" />
+
+The authors mention this programming model has several limitations. For instance, it cannot support recursive calls, multi-dimensional arrays, and non-loop-variable indices. These limitations are reasonable. Even for nowadays HLS tools, they still cannot support recursive function calls[^2].
+
+#### Partition Type
+> "Partitioning a program can be thought of as a type inference on the partition types."
+
+I think this is the most interesting part of the paper. The authors add partition information to the type system, thus partitioning the program can be automatically done using type inference. The definition of partition type is shown below.
+
+<center>
+<img src="partition-inference.png" alt="alt_text" title="image_tooltip" style="zoom:40%;" />
+</center>
+
+#### Partition Process
+The partition process in the paper is a bit messy, so I reorganize it in a more clear way. We have the following steps:
```

Comment:
Also super helpful to use your own organization instead of the one in the paper, if you find it easier to ingest that way! Nice work; this is very clear.